runtime/lua/vim/lsp/_folding_range.lua

```diff
@@ -0,0 +1,404 @@
+end
+
+-- The "garbage collector" for `bufstate` and its associated event hooks.
+-- See the comment above `augroup_setup` function for more details.
+--
+-- It will be fine without this, as the bufstate and event hooks will also
+-- be cleared when the buffer is unloaded.
+-- api.nvim_create_autocmd('OptionSet', {
+--   group = augroup_setup,
+--   pattern = 'foldexpr',
+--   callback = function()
+--     -- Sould find a more reliable way to detect changes in `foldexpr`.
+--     if vim.v.option_old == 'v:lua.vim.lsp.foldexpr()' then
+--       for bufnr in pairs(bufstates) do
+--         if
+--           ---@param winid integer
+--           vim.iter(vim.fn.win_findbuf(bufnr)):all(function(winid)
+--             return vim.wo[winid].foldexpr ~= 'v:lua.vim.lsp.foldexpr()'
+--           end)
+--         then
+--           bufstates[bufnr] = nil
+--           api.nvim_clear_autocmds({ buffer = bufnr, group = augroup_setup })
+--         end
+--       end
+--     end
+--   end,
+-- })
+
+---@param kind lsp.FoldingRangeKind
+---@param winid integer
+local function foldclose(kind, winid)
+  vim._with({ win = winid }, function()
+    local bufnr = api.nvim_win_get_buf(winid)
+    local row_kinds = bufstates[bufnr].row_kinds
+    -- Reverse traverse to ensure that the smallest ranges are closed first.
+    for row = api.nvim_buf_line_count(bufnr) - 1, 0, -1 do
+      local kinds = row_kinds[row]
+      if kinds and kinds[kind] then
+        vim.cmd(row + 1 .. 'foldclose')
+      end
+    end
+  end)
+end
+
+---@param kind lsp.FoldingRangeKind Kind to close, one of "comment", "imports" or "region".
+---@param winid? integer Defaults to the current window.
+function M.foldclose(kind, winid)
+  vim.validate('kind', kind, 'string')
+  vim.validate('winid', winid, 'number', true)
+
+  winid = winid or api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winid)
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return
+  end
+
+  if bufstate.version == util.buf_versions[bufnr] then
+    foldclose(kind, winid)
+    return
+  end
+  -- Schedule `foldclose()` if the buffer is not up-to-date.
+
+  --- Index in the form of request_id -> true?
+  ---@type table<integer, true?>
+  local scheduled_request = {}
+  for _, client in
+    ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+  do
+    local request_id = request(bufnr, client)
+    if request_id then
+      scheduled_request[request_id] = true
+    end
+  end
+
+  api.nvim_create_autocmd('LspRequest', {
+    buffer = bufnr,
+    callback = function(args)
+      ---@type integer
+      local request_id = args.data.request_id
+      if scheduled_request[request_id] and args.data.request.type == 'complete' then
+        scheduled_request[request_id] = nil
+      end
+      -- Do `foldclose()` if all the requests is completed.
+      if next(scheduled_request) == nil then
+        foldclose(kind, winid)
+        return true
+      end
+    end,
+  })
```

Comment:
A bit an unusual pattern to use `LspRequest` for a countdown latch to react once all responses were received.
Typical pattern is to use a counter like in `buf_request_all` or in some of the methods in lsp.buf.

But no strong opinion on this.

Reply:
Hmm, yeah I think that might be better.

Reply:
Okay, I will add another `lsp.MultiHandler` and reduce the number of calls to `foldupdate`.